feat: Manage org read-only state + superadmin view refresh

SuaYoo opened the pull request:
Resolves https://github.com/webrecorder/browsertrix/issues/1903

Note: This was branched from https://github.com/webrecorder/browsertrix/pull/1788 to get ahead of potential conflicts

### Changes

- Displays org status in superadmin view
- Enables superadmins to update org read-only status
- Sorts orgs alphabetically
- Refactors org list to `TailwindComponent`

### Manual testing

1. Log in as superadmin
2. Click "..." action menu for an org
3. Click "Make Read-Only". Verify dialog is shown
4. Enter reason and submit. Verify org is now read-only
5. Click action menu > "Undo Read-Only". Verify org is no longer read-only

### Screenshots

| Page | Image/video |
| ---- | ----------- |
| Superadmin Dashboard | <img width="1026" alt="Screenshot 2024-07-01 at 10 13 21 PM" src="https://github.com/webrecorder/browsertrix/assets/4672952/0ac304fb-bdc7-405d-af55-9264b3c1fe1f"> |
| Superadmin Dashboard | <img width="601" alt="Screenshot 2024-07-01 at 10 13 28 PM" src="https://github.com/webrecorder/browsertrix/assets/4672952/ad80bff1-ca87-4296-9bf8-175df61c3009"> |
| Superadmin Dashboard | <img width="1034" alt="Screenshot 2024-07-01 at 10 13 42 PM" src="https://github.com/webrecorder/browsertrix/assets/4672952/4794688a-d408-4500-85c6-5aa974568943">

<!-- ### Follow-ups -->

SuaYoo committed:
fix dialog

SuaYoo committed:
add dialog

SuaYoo committed:
check confirmation string

SuaYoo committed:
sort orgs alphabetically

SuaYoo committed:
refactor org list

SuaYoo committed:
show additional data

SuaYoo committed:
show quota status

SuaYoo committed:
update layout

SuaYoo committed:
make read only

emma-sg commented:
Could we use a datalist here rather than a text input to suggest preset options rather than just freeform text?

SuaYoo committed:
fix tests

tw4l reviewed (approved):
Might be nice to always have the default org display first in the Organizations list? Though I can see an argument for straight alphabetical as well. Otherwise this looks great!

SuaYoo commented:
> Could we use a datalist here rather than a text input to suggest preset options rather than just freeform text?

That makes sense--what are the options?

SuaYoo commented:
> Might be nice to always have the default org display first in the Organizations list? Though I can see an argument for straight alphabetical as well. Otherwise this looks great!

Could we address this with backend sorting? We'll want that pretty soon anyway, so that the list can be paginated and sorted.

Shrinks99 commented on frontend/src/components/orgs-list.ts:
```diff
@@ -110,63 +225,204 @@ export class OrgsList extends LiteElement {
+    let status = {
+      icon: html`<sl-icon
+        class="text-base text-success"
+        name="check-circle"
```
We don't use the non filled variant for statuses anymore with the exception of `slash-circle` to give it a more unique shape from the rest.

```suggestion
        name="check-circle-fill"
```

tw4l committed:
Add and apply backend sorting for org list

The default org will always be sorted first, regardless of sort options.
Orgs after the first will be sorted by name ascending by default.
Sorting currently supported on name, slug, and readOnly.

tw4l commented on frontend/src/components/orgs-list.ts:
```diff
@@ -93,6 +144,70 @@ export class OrgsList extends LiteElement {
+                html`Are you sure you want to make
+                  <span class="font-semibold">${org.name}</span>
+                  read-only? All members' access will be reduced to the viewer
+                  role.`,
```
Hm, not quite true anymore, as when an org is read-only we're now going to leave user permissions as they are and they'll still be able to see e.g. QA results.

Maybe "Are you sure you want to make ORG read-only? Members will no longer be able to crawl, upload files, create browser profiles, or create collections." ?

tw4l committed:
Fix linting issues

tw4l committed:
Add one more pylint disable comment

tw4l committed:
Merge branch 'main' into frontend-org-manage-readonly

tw4l commented:
> > Could we use a datalist here rather than a text input to suggest preset options rather than just freeform text?
> 
> That makes sense--what are the options?

I might suggest leaving it open, at least for now or until we have brainstormed what those options should be?

tw4l committed:
Fix typing

tw4l commented:
> > Might be nice to always have the default org display first in the Organizations list? Though I can see an argument for straight alphabetical as well. Otherwise this looks great!
> 
> Could we address this with backend sorting? We'll want that pretty soon anyway, so that the list can be paginated and sorted.

Done! Default org is always sorted first, otherwise defaults to alphabetical ascending with support for `slug` and `readOnly` as sortBy values as well.

Shrinks99 reviewed (changes requested):
Seems to work as expected, nice!

If possible could we set all the relevant actions to a disabled button state?  The current implementation allows people to progress to stages that they can't complete which isn't great.  Users shouldn't be able to enter the upload dialog if they don't have permission to actually upload anything for example.  I don't think we need to give these states individual tooltips as this will be coupled with a global banner explaining why actions can't be accessed in the future.

Off the top of my head:
- The workflow creator and editor _should_ work, but the "run on save" toggle should be switched to off and disabled
- Upload button should be disabled
- New browser profile button should be disabled
  - Currently users can go through the whole browser creation process, only to not be allowed to save it!
- Edit browser profile button should be disabled
- Run crawl should be disabled
- Creating and editing collections should maybe also be disabled for consistency?
  - Sharing settings should remain available.

ikreymer commented:
> Seems to work as expected, nice!
> 
> If possible could we set all the relevant actions to a disabled button state? The current implementation allows people to progress to stages that they can't complete which isn't great. Users shouldn't be able to enter the upload dialog if they don't have permission to actually upload anything for example. I don't think we need to give these states individual tooltips as this will be coupled with a global banner explaining why actions can't be accessed in the future.
> 
> Off the top of my head:
> 
> * The workflow creator and editor _should_ work, but the "run on save" toggle should be switched to off and disabled
> * Upload button should be disabled
> * New browser profile button should be disabled
>   
>   * Currently users can go through the whole browser creation process, only to not be allowed to save it!
> * Edit browser profile button should be disabled
> * Run crawl should be disabled
> * Creating and editing collections should maybe also be disabled for consistency?
>   
>   * Sharing settings should remain available.

This is a lot of additional work for us, without much benefit. User sees the read-only state and is notified they can't do the action if they try. The read-only state is designed as a temporary error state to cure some other issue and not a typical state, so these things should be fine for now.

Shrinks99 commented:
> User sees the read-only state and is notified they can't do the action if they try.  The read-only state is designed as a temporary error state to cure some other issue

I think you've succinctly explained why this actually would be a valuable change.  Disabling these controls reduces a user's chances of attempting to accomplish invalid actions and increases the chance that they will take the correct path and attempt to cure the state.  This reduces complexity for our customers and is in-line with our philosophy of hiding controls that viewers don't have access to.

Just because ideally this state isn't triggered and hopefully few users are in this state at any given time doesn't mean it should function poorly.  Letting users access controls that only tell them they can't accomplish the task they were trying to at the _end_ of the process is a bad user experience that, if I'm not mistaken, can be solved with some extra boolean logic checking for if OrgData:`readOnly` is true?

tw4l commented:
> > User sees the read-only state and is notified they can't do the action if they try.  The read-only state is designed as a temporary error state to cure some other issue
> 
> I think you've succinctly explained why this actually would be a valuable change. Disabling these controls reduces a user's chances of attempting to accomplish invalid actions and increases the chance that they will take the correct path and attempt to cure the state. This reduces complexity for our customers and is in-line with our philosophy of hiding controls that viewers don't have access to.
> 
> Just because ideally this state isn't triggered and hopefully few users are in this state at any given time doesn't mean it should function poorly. Letting users access controls that only tell them they can't accomplish the task they were trying to at the _end_ of the process is a bad user experience that, if I'm not mistaken, can be solved with some extra boolean logic checking for if OrgData:`readOnly` is true?

I'd agree that this seems worthwhile, should just be changing the disabled logic for a handful of buttons, I don't think that would take too long? And would make this more consistent with how storage quotas are handled.

ikreymer commented:
> I'd agree that this seems worthwhile, should just be changing the disabled logic for a handful of buttons, I don't think that would take too long? And would make this more consistent with how storage quotas are handled.

Ah ok, sure, if we do this for org quotas already, then that does make sense for consistency.

emma-sg commented:
I guess an ideal solution here might also include checking for read-only states on the backend in more places too — but I know this was primarily intended to be a front-end PR

SuaYoo committed:
Update frontend/src/components/orgs-list.ts

Co-authored-by: Henry Wilkinson <henry@wilkinson.graphics>

SuaYoo commented:
> If possible could we set all the relevant actions to a disabled button state?

I spun this out into a separate issue since the scope of this PR is superadmin changes.

SuaYoo committed:
update copy

SuaYoo committed:
Merge branch 'main' into frontend-org-manage-readonly
